### ISSUE TITLE
Malformed response triggers uncatchable error, crashing the calling program

### ISSUE THREAD

**Issue by phretor** [mcollina/node-coap#168]
If the option length is > 16 bits, the following line throws an exception:

https://github.com/mcollina/node-coap/blob/master/lib/option_converter.js#L140

```
var contentFormatToString = function(value) {
  if (value.length === 0)
    return formatsBinaries[0]

  if (value.length === 1)
    value = value.readUInt8(0)
  else if (value.length === 2)
    value = value.readUInt16BE(0)
  else
    throw new Error('Content-Format option is too big')

  var result = formatsBinaries[value]

  if (!result)
    throw new Error('No matching format found')

  return result
}
```

Because of how `contentFormatToString` is called, I can't find a way to catch that error in the calling program.

If I understand the RFC correctly, the max length is indeed 2 bytes, so the error is worth throwing. However, there must be a way to properly allow the calling program to catch it (e.g., and decide to skip the packet).

I'm proposing the following ugly patch in `incoming_message.js`:

```
function IncomingMessage(packet, rsinfo, outSocket) {
  Readable.call(this)

  try {
    pktToMsg(this, packet)
  } catch (e) {
    //any idea on how to emit a proper event here?
  }

  this.rsinfo = rsinfo
  this.outSocket = outSocket

  this._packet = packet
  this._payloadIndex = 0
}
```

Am I overlooking something?

**Comment by mkovatsc** [comment]
Applications are also forced to crash when a server sends an unknown/unregistered Content-Format number, as the error rooted in `UDP.onMessage` cannot be caught.

Also need proper error event for this, so it does not force-crash the application.

**Comment by GiedriusM** [comment]
IMO any incoming packet handling exceptions should be caught and ignored. What do you guys think?

**Comment by mkovatsc** [comment]
I was thinking about an error handler that could be defined on Agent to see errors and if it is not defined, then maybe just ignore them. However, then there will be open requests that never receive a response, e.g., because the application did not know about the Content-Format beforehand (which would be totally fine for generic clients). So there should also be an error handler on the request message to deal with that.

**Comment by neophob** [comment]
error handler sounds good to me

**Comment by mkovatsc** [comment]
I currently only have the client side of the issue. More thoughts are needed for server-side incoming errors that originate in UDP.onMessage.

For all socket-level errors for clients, the Agent is probably the best place to catch. 

For message-level errors (i.e., no useful response IncomingMessage can be created), the OutgoingMessage (i.e., request) should be notified on its `error` handler.

**Comment by phretor** [comment]
@mkovatsc check if #194 works for you. It's a quick (and very dirty, because I'm not taking care of the agent state) fix, but at least helps me catching a simple error that would have otherwise crashed the entire process.

**Comment by phretor** [comment]
@mkovatsc I think at this point is easier to just `return value` in stead of `throw new Error('Content-Format option is too big')` leaving it up to the application to do the validation as you suggest.

Thanks!

**Comment by JKRhb** [comment]
This issue should now be fixed with the merge of PR #196. Please let us know if the issue should persist in any way :)
